Mathlib/Topology/Order/ScottTopology.lean

```diff
@@ -0,0 +1,439 @@
+      exact image_inter_nonempty_iff.mp $ hu.2 _ _ (hd₁.image f)
+          (directedOn_image.mpr (hd₂.mono (by simp only [Order.Preimage]; apply h.monotone)))
+          (h hd₁ hd₂ hd₃) ha
+  · intros hf d d₁ d₂ a d₃
+    rw [IsLUB]
+    constructor
+    · apply Monotone.mem_upperBounds_image (continuous_monotone hf) ((isLUB_le_iff d₃).mp le_rfl)
+    · rw [lowerBounds, mem_setOf_eq]
+      intros b hb
+      let u := (Iic b)ᶜ
+      by_contra h
+      have s1 : IsOpen u := by
+        rw [isOpen_compl_iff, ← closure_singleton]
+        exact isClosed_closure
+      have s2 : IsOpen (f⁻¹'  u) := IsOpen.preimage hf s1
+      rw [isOpen_iff_upper_and_inaccessible_by_directed_joins] at s2
+      obtain ⟨c, h_1_left, h_1_right⟩ := s2.2 d a d₁ d₂ d₃ h
+      simp at h_1_right
+      rw [upperBounds] at hb
+      simp at hb
+      have c1: f c ≤ b := hb _ h_1_left
+      contradiction
+
+end preorder
+
+section partial_order
+variable [PartialOrder α] [TopologicalSpace α] [ScottTopology α]
+
+/--
+The Scott topology on a partial order is T₀.
+-/
+-- see Note [lower instance priority]
+instance (priority := 90): T0Space α :=
+(t0Space_iff_inseparable α).2 $ fun x y h => Iic_injective $
+  by simpa only [inseparable_iff_closure_eq, ScottTopology.closure_singleton] using h
+
+end partial_order
+
+end ScottTopology
+
+section complete_lattice
+
+variable [CompleteLattice α] [TopologicalSpace α] [ScottTopology α]
+
+lemma isOpen_iff_isUpperSet_and_sup_mem_implies_tail_subset {u : Set α} :
+  IsOpen u ↔ (IsUpperSet u ∧ ∀ ⦃d : Set α⦄,
+    d.Nonempty → DirectedOn (· ≤ ·) d → sSup d ∈ u → ∃ b ∈ d, Ici b ∩ d ⊆ u) := by
+  rw [ScottTopology.isOpen_iff_upper_and_Scott_Hausdorff_Open]
+  apply and_congr_right'
+  constructor
+  · exact fun h d hd₁ hd₂ hd₃ => h d (sSup d) hd₁ hd₂ (isLUB_sSup d) hd₃
+  · exact fun h d a hd₁ hd₂ hd₃ ha => h hd₁ hd₂ (Set.mem_of_eq_of_mem (IsLUB.sSup_eq hd₃) ha)
+
+lemma isOpen_iff_upper_and_sup_mem_implies_inter_nonempty
+{u : Set α} : IsOpen u ↔
+(IsUpperSet u ∧  ∀ (d : Set α), d.Nonempty → DirectedOn (· ≤ ·) d → sSup d ∈ u →
+(d∩u).Nonempty) := by
+  rw [ScottTopology.isOpen_iff_upper_and_inaccessible_by_directed_joins]
+  apply and_congr_right'
+  constructor
+  · exact fun h d hd₁ hd₂ hd₃ => h d (sSup d) hd₁ hd₂ (isLUB_sSup d) hd₃
+  · exact fun h d a hd₁ hd₂ hd₃ ha => h d hd₁ hd₂ (Set.mem_of_eq_of_mem (IsLUB.sSup_eq hd₃) ha)
+
+lemma isClosed_iff_lower_and_closed_under_Directed_Sup {s : Set α} : IsClosed s
+  ↔ (IsLowerSet s ∧
+  ∀ (d : Set α), d.Nonempty → DirectedOn (· ≤ ·) d → d ⊆ s → sSup d ∈ s ) := by
+  rw [ScottTopology.isClosed_iff_lower_and_subset_implies_LUB_mem]
+  apply and_congr_right'
+  constructor
+  · exact fun h d hd₁ hd₂ hd₃ => h d (sSup d) hd₁ hd₂ (isLUB_sSup d) hd₃
+  · exact fun h d a h₁ h₂ h₃ ha => (Set.mem_of_eq_of_mem (IsLUB.sSup_eq h₃).symm (h d h₁ h₂ ha))
+
+
+end complete_lattice
+
+variable [Preorder α]
+
+lemma Scott_Hausdorff_le_Scott [TopologicalSpace α] [ScottTopology α] :
```

Comment:
```suggestion
lemma scottHausdorffTopology_le_of_scottTopology [TopologicalSpace α] [ScottTopology α] :
```
You might want to declare your predicates and topologies in the `Topology` namespace to avoid those stupidly long lemma names.

Reply:
Do you mean `def Topology.ScottHausdorff : TopologicalSpace α :=...`? I don't see how that would change the length of the lemma names unless they were also in the Topology namespace?